### PR TITLE
Adds a way of merging models with different sizes(B) 

### PR DESCRIPTION
**Pull request by Ar57m**
I'm adding the possibility of merging models with different amounts of parameters(Bs) which have the same amount of layers, through task arithmetic. I kinda hardcoded generalized task arithmetic to make it work, it probably needs more testing to see if it's useful(or it's working properly) this approach, I currently don't have the skills and the way to properly test it. Edit or refactore it and implement as your wish.

### CODE REVIEW

**Comment by Ar57m** [comment]
before I tried slicing the tensors to match the small base model size, but it seems very unstable even at low weights(like 0.01, 0.005), so in this branch breaking_math I used interpolation to kinda scale up or down the tensors, it seems much more stable. I don't know if it's just noise that's being added this way(scientific testing needed).
Some trials using this method:

base: erfanzar/LinguaMatic-2.7B + PygmalionAI/pygmalion-2-7b  weight: 0.01
https://huggingface.co/Aryanne/pyg2broken/tree/main


base: erfanzar/LinguaMatic-2.7B + PygmalionAI/pygmalion-2-7b  weight: 0.1
https://huggingface.co/Aryanne/pyg2broken/tree/pyg_at_10_percent


base: erfanzar/LinguaMatic-2.7B + PygmalionAI/pygmalion-2-7b  weight: 1.0
https://huggingface.co/Aryanne/pyg2broken/tree/759a7eae6594616e703c24d94821931c2a320728

this last is the most unstable, produces a lot of gibberish.

The model with weight 0.1 I couldn't upload the safetensors because of network issues, so just I uploaded the quants. The model with weight 1.0 I couldn't upload the quants too.

**Comment by cg123** [comment]
This is really interesting! I'll definitely have to play with it a bit.

I'm a little surprised this kind of interpolation isn't more immediately harmful to the models. Will be interesting to see if it's just adding noise or if there's something spicier going on.

Thanks for the PR!

**Comment by Ar57m** [comment]
> This is really interesting! I'll definitely have to play with it a bit.
> 
> I'm a little surprised this kind of interpolation isn't more immediately harmful to the models. Will be interesting to see if it's just adding noise or if there's something spicier going on.
> 
> Thanks for the PR!

Noticed now I was probably using the wrong mode(nearest)  to interpolate, it was basically cutting the bigger model to the size of the smaller not actually scaling down, the right mode I think is bilinear and linear. Gonna change now

**Comment by Ar57m** [comment]
I did Sheared-LLaMA-2.7B-ShareGPT + SanjiWatsuki/Silicon-Maid-7B at 10% if you guys wanna test [Aryanne/sheared-silicon10p](https://huggingface.co/Aryanne/sheared-silicon10p), if I did everything right(and didn't break anything), this should be working better than before.

**Comment by Ar57m** [comment]
here's a code that shows what bilinear/linear is doing:
```
import torch
import torch.nn.functional as F
torch.manual_seed(42)

x = torch.tensor([[1., 3.],[5., 7.],[9.,11.],[13.,15.]])

y = torch.tensor([[2.0, 4.0],[6.0,8.0],[10.,12.]])


print("X :")
print(x)

print("\nBase :")
print(y)

if x.device.type == "cpu":
   x = x.to(torch.float32)
if x.shape != y.shape:
   if x.ndim == 2 and (x.shape[0] != y.shape[0] or x.shape[1] != y.shape[1]):
      x = F.interpolate(x.unsqueeze(0).unsqueeze(0), size=(y.shape[0],y.shape[1]), mode='bilinear').squeeze()/2
   if x.ndim == 1 and x.shape[0] != y.shape[0] :
      x = F.interpolate(x.unsqueeze(0).unsqueeze(0), size=y.shape, mode='linear' ).squeeze()
   x = x.to(y.dtype)

print("X after interpolation:")
print(x)
```

The output:

```
X :
tensor([[ 1.,  3.],
        [ 5.,  7.],
        [ 9., 11.],
        [13., 15.]])

Base :
tensor([[ 2.,  4.],
        [ 6.,  8.],
        [10., 12.]])
X after interpolation:
tensor([[ 1.6667,  3.6667],
        [ 7.0000,  9.0000],
        [12.3333, 14.3333]])
```

**Review by winlover32 — APPROVED** [review]

**Review by winlover32 — APPROVED** [review]
check

**Comment by Ar57m** [comment]
> check

thanks for the check, this branch seems to be degrading a lot the base model(I submitted some trials using this branch and the other weighting_interp to the open-llm-leaderboard, search "Aryanne;sheared" to compare), at the moment I'm messing/testing some things on this [branch](https://github.com/Ar57m/mergekit/tree/weighting_interp) to see if I can improve the base model or pass some characteristics, currently appears to be just noise.

**Comment by Ar57m** [comment]
idk, but this doesn't seems very promising at moment, don't know if there's a good way of merging models with different sizes.

btw @cg123 , I've been messing with other new(I think) method [here](https://github.com/Ar57m/mergekit/tree/swapping) ,  which takes the base and uses a chess grid mask(at the value level) to put one model on it. if you have time to take a look.

**Comment by david565656** [comment]
IS this in the main branch? avail?
This can solve an issue merging 7b and 13b models.
As well as 10.7/11b and 13b models too.

**Comment by Ar57m** [comment]
@david565656 hey, this pr I was trying in the breaking_math and weighting_interp branch.

[here](https://huggingface.co/Aryanne/9b_plus_tie13/tree/main) I did a merge a 13b into a 9b, but probably made the model invent words, did a bunch of merges 7b into a 3b, it was generating gibberish in most cases, I closed this pr cause the merges was more breaking the models than doing something useful.

I don't remember if I tried padding, I was interpolating in both branches which seems to break too much.

**Comment by david565656** [comment]
Thank you  ; I will try this out - got the two files and will give it a go.
My case is the opposite -> Expanding a model.

First -> Mistral 0.2 -> 13B ; then use it as a base to merge 13B models onto to try and get true 32k context.
Eventually boost Mistral 0.2 -> 20B ... then merge some great 20B models with it... 32k context.

When I try to "merge" a 7B with a 13B -> Mergekit errors out do to mismatch dimensions (??) of the matrices.
Also looked at a lot of merges online and the use of a "healing process" post "Upscale" seems to work. (still testing this).
Found this in a 3 step merge a few days ago.

This might or might not fix "gibberish" issues...
Another method I want to try is creating masks of upscale models to then fix a merged+upscale model as a "light" merge to fix the merge/scaled.
IE -> Upscale (keeping order, no interlacing) -> Use as a base + merge other model(s) [also upscaled - no interlace] -> then "remerge" the upscale from step 1 "lightly" with this model to repair/smooth over some damage.

I have made some strange upscale/frankinsteins - that work despite no "smoothing" end process. They can go gibberish y at times, but at other times the creative output is astonishing... and after testing over 550 models to date... that is a find.

Will let you know what happens with your scripts shortly... thanks again for all you do.
